File: examples/bad/typescript/posts-router-bad.ts
// BAD EXAMPLE: Do NOT copy this style.

import express from 'express';
import { Pool } from 'pg';

const router = express.Router();
const pool = new Pool({ connectionString: process.env.DATABASE_URL });

router.post('/doCreatePost', async (req, res) => {
  // No auth, no validation, raw SQL string concatenation.
  const { title, body, userId } = req.body;

  const result = await pool.query(
    `INSERT INTO posts (title, body, user_id) VALUES ('${title}', '${body}', ${userId}) RETURNING *`,
  );

  res.send(result.rows[0]); // raw row, wrong status code
});

router.get('/getAllPosts', async (_req, res) => {
  const result = await pool.query('SELECT * FROM posts');
  res.send(result.rows); // leaking internal schema
});

export default router;

File: examples/bad/python/posts-router-bad.py
# BAD EXAMPLE: Do NOT copy this style.

from fastapi import FastAPI, Request
import sqlite3

app = FastAPI()
conn = sqlite3.connect('blog.db')


@app.post('/createPost')
async def create_post(request: Request):
    data = await request.json()
    # No validation, SQL injection risk
    conn.execute(
        f"INSERT INTO posts (title, body, user_id) VALUES ('{data['title']}', '{data['body']}', {data['user_id']})"
    )
    conn.commit()
    return {'status': 'ok'}  # no id, no status code handling


@app.get('/getPosts')
async def get_posts():
    cur = conn.cursor()
    cur.execute('SELECT * FROM posts')
    return cur.fetchall()  # raw tuples
